src/interaction/sessionrecording.cpp

```diff
@@ -0,0 +1,1207 @@
+    _idxTimeline_cameraPtrNext = 0;
+    _idxTimeline_cameraPtrPrev = 0;
+    _hasHitEndOfCameraKeyframes = false;
+
+    _cleanupNeeded = false;
+}
+
+bool SessionRecording::isDataModeBinary() {
+    if (_recordingDataMode == recordDataMode::binary)
+        return true;
+    else
+        return false;
+}
+
+void SessionRecording::writeToFileBuffer(const double& src) {
+    const size_t writeSize_bytes = sizeof(double);
+    unsigned char const *p = reinterpret_cast<unsigned char const*>(&src);
+    memcpy((_keyframeBuffer + _bufferIndex), p, writeSize_bytes);
+    _bufferIndex += writeSize_bytes;
+}
+
+void SessionRecording::writeToFileBuffer(std::vector<char>& cvec) {
+    const size_t writeSize_bytes = cvec.size() * sizeof(char);
+    memcpy((_keyframeBuffer + _bufferIndex), &cvec[0], writeSize_bytes);
+    _bufferIndex += writeSize_bytes;
+}
+
+void SessionRecording::writeToFileBuffer(const unsigned char c) {
+    const size_t writeSize_bytes = sizeof(char);
+    _keyframeBuffer[_bufferIndex] = c;
+    _bufferIndex += writeSize_bytes;
+}
+
+void SessionRecording::writeToFileBuffer(const bool b) {
+    const size_t writeSize_bytes = sizeof(char);
+    if( b )
+        _keyframeBuffer[_bufferIndex] = 1;
+    else
+        _keyframeBuffer[_bufferIndex] = 0;
+    _bufferIndex += writeSize_bytes;
+}
+
+void SessionRecording::saveStringToFile(const std::string s) {
+    size_t strLen = s.size();
+    size_t writeSize_bytes = sizeof(size_t);
+
+    _bufferIndex = 0;
+    unsigned char const *p = reinterpret_cast<unsigned char const*>(&strLen);
+    memcpy((_keyframeBuffer + _bufferIndex), p, writeSize_bytes);
+    _bufferIndex += (unsigned int)writeSize_bytes;
+    saveKeyframeToFileBinary(_keyframeBuffer, _bufferIndex);
+
+    _recordFile.write(s.c_str(), s.size());
+}
+
+void SessionRecording::readFromPlayback(unsigned char& result) {
+    _playbackFile.read(reinterpret_cast<char*>(&result), sizeof(unsigned char));
+}
+
+void SessionRecording::readFromPlayback(double& result) {
+    _playbackFile.read(reinterpret_cast<char*>(&result), sizeof(double));
+}
+
+void SessionRecording::readFromPlayback(float& result) {
+    _playbackFile.read(reinterpret_cast<char*>(&result), sizeof(float));
+}
+
+void SessionRecording::readFromPlayback(size_t& result) {
+    _playbackFile.read(reinterpret_cast<char*>(&result), sizeof(size_t));
+}
+
+void SessionRecording::readFromPlayback(bool& result) {
+    unsigned char b;
+    _playbackFile.read(reinterpret_cast<char*>(&b), sizeof(unsigned char));
+    if (b == 0)
+        result = false;
+    else if (b == 1)
+        result = true;
+    else
+        LERROR(fmt::format("Invalid bool read value at {}", _playbackLineNum - 1));
+}
+
+void SessionRecording::readFromPlayback(std::string& result) {
+    result.erase();
+    size_t strLen;
+    //Read string length from file
+    _playbackFile.read(reinterpret_cast<char*>(&strLen), sizeof(strLen));
+    //Read back full string
+    char* temp = new char[strLen + 1];
```

Comment:
Better to use `std::vector<char>(strLen + 1)`

Reply:
Changed